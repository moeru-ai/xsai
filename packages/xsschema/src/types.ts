import type { StandardSchemaV1 } from '@standard-schema/spec'

export type { StandardSchemaV1 as Schema } from '@standard-schema/spec'
export type { JSONSchema7 as JSONSchema } from 'json-schema'

export type Infer<T extends StandardSchemaV1> = StandardSchemaV1.InferOutput<T>
export type InferIn<T extends StandardSchemaV1> = StandardSchemaV1.InferInput<T>
