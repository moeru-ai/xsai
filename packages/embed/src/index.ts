export { embed as default, embed, type EmbedOptions, type EmbedResponseUsage, type EmbedResult } from './utils/embed'
export { embedMany, type EmbedManyOptions, type EmbedManyResult } from './utils/embed-many'
