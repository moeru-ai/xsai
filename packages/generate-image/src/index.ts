import type { CommonRequestOptions } from '@xsai/shared'

import { requestBody, requestHeaders, requestURL, responseCatch, responseJSON } from '@xsai/shared'

export interface GenerateImageOptions extends CommonRequestOptions {
  [key: string]: unknown
  /**
   * The number of images to generate.
   * @default `1`
   */
  n?: number
  /** A text description of the desired image(s). */
  prompt: string
  /**
   * The format in which generated images with `dall-e-2` and `dall-e-3` are returned from the API.
   * `gpt-image-1` only supports `b64_json`, and `url` will be ignored.
   *
   * Images will always be Base64-encoded in the result. If `url` is used, the images will be
   * fetched upon receiving the response.
   */
  responseFormat?: 'b64_json' | 'url'
  /**
   * The size of the generated images.
   * @default `1024x1024`
   */
  size?: `${number}x${number}`
}

export interface GenerateImageResponse {
  created: number
  data: {
    /**
     * The Bse64-encoded JSON of the generated image. Default value for `gpt-image-1`, and only
     * present if `response_format` is set to `b64_json` for `dall-e-2` and `dall-e-3`.
     */
    b64_json?: string
    /**
     * For `dall-e-3` only, the revised prompt that was used to generate the image.
     */
    revised_prompt?: string
    /**
     * When using `dall-e-2` or `dall-e-3`, the URL of the generated image if `response_format` is
     * set to `url` (default value). Unsupported for `gpt-image-1`.
     */
    url?: string
  }[]
}

export interface GenerateImageResult {
  image: GenerateImageResultImage
  images: GenerateImageResultImage[]
}

export interface GenerateImageResultImage {
  base64: string
  mimeType: string
}

/** @internal */
const mimeTypes = {
  '/9j/': 'image/jpg',
  'AAAAIGZ0eXBhdmlm': 'image/avif',
  'iVBORw0KGgo': 'image/png',
  'R0lGOD': 'image/gif',
  'UklGRg==': 'image/webp',
}

/** @internal */
const convertImage = (b64_json: string) => {
  const key = Object.keys(mimeTypes).find(prefix => b64_json.startsWith(prefix)) as keyof typeof mimeTypes | undefined
  const mimeType = mimeTypes[key ?? 'iVBORw0KGgo']

  return {
    // eslint-disable-next-line @masknet/string-no-data-url
    base64: `data:${mimeType};base64,${b64_json}`,
    mimeType,
  }
}

/** @internal */
const responseBlobAsDataURL = async (res: Response): Promise<string> =>
  responseCatch(res)
    .then(async (res) => {
      const blob = await res.blob()

      try {
        return await new Promise<string>((resolve, reject) => {
          const reader = new FileReader()
          reader.onloadend = () => resolve(reader.result as string)
          reader.onerror = reject
          reader.readAsDataURL(blob)
        })
      }
      catch {
        throw new Error(`Failed to parse response blob, response URL: ${res.url}`)
      }
    })

/** @experimental */
export const generateImage = async (options: GenerateImageOptions): Promise<GenerateImageResult> =>
  (options.fetch ?? globalThis.fetch)(requestURL('images/generations', options.baseURL), {
    body: requestBody(options),
    headers: requestHeaders({
      'Content-Type': 'application/json',
      ...options.headers,
    }, options.apiKey),
    method: 'POST',
    signal: options.abortSignal,
  })
    .then(responseJSON<GenerateImageResponse>)
    .then(async ({ data }) =>
      Promise.all(
        data.map(async (img, i) => {
          if (typeof img.b64_json === 'string') {
            return convertImage(img.b64_json)
          }

          if (typeof img.url === 'string') {
            return (options.fetch ?? globalThis.fetch)(new URL(img.url), {
              signal: options.abortSignal,
            })
              .then(responseBlobAsDataURL)
              .then((dataURL) => {
                const sepIndex = dataURL.indexOf(';')
                const mimeType = dataURL.substring(5, sepIndex) // `data:(...);...`
                return {
                  base64: dataURL,
                  mimeType,
                }
              })
          }

          throw new Error(`Unrecognized image at index ${i}: ${JSON.stringify(img)}`)
        }),
      ),
    )
    .then(images => ({ image: images[0], images }))
