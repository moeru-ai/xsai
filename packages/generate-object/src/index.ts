import { type Infer, type Schema, toJSONSchema, validate } from '@typeschema/main'
import { generateText, type GenerateTextOptions, type GenerateTextResult } from '@xsai/generate-text'
import { clean } from '@xsai/shared'

export interface GenerateObjectOptions<T extends Schema> extends GenerateTextOptions {
  schema: T
  schemaDescription?: string
  schemaName?: string
}

// TODO: toolCalls, toolResults
export interface GenerateObjectResult<T extends Schema> extends Omit<GenerateTextResult, 'text' | 'toolCalls' | 'toolResults'> {
  object: Infer<T>
}

/** @experimental WIP */
export const generateObject = async <T extends Schema>(options: GenerateObjectOptions<T>): Promise<GenerateObjectResult<T>> =>
  await generateText({
    ...options,
    response_format: {
      json_schema: {
        description: options.schemaDescription,
        name: options.schemaName ?? 'json_schema',
        schema: await toJSONSchema(options.schema)
          .then(json => clean({
            ...json,
            $schema: undefined,
          })),
        strict: true,
      },
      type: 'json_schema',
    },
    schema: undefined,
    schemaDescription: undefined,
    schemaName: undefined,
  })
    .then(async ({ finishReason, steps, text, usage }) => {
      const result = await validate(options.schema, JSON.parse(text!))

      if (result.success) {
        return {
          finishReason,
          object: result.data,
          steps,
          usage,
        }
      }
      else {
        throw new Error([
          'Schema validation failed:',
          ...result.issues.map(issue => `- ${issue.message}`),
        ].join('\n'))
      }
    })

export default generateObject
