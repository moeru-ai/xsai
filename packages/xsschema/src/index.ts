export { toJsonSchema, toJsonSchema as toJSONSchema } from './to-json-schema'
export type { Infer, InferIn, JSONSchema as JsonSchema, JSONSchema, Schema } from './types'
export { validate } from './validate'
