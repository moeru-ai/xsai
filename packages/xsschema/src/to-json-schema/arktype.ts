import type { Type } from 'arktype'
import type { JSONSchema7 } from 'json-schema'

export const toJsonSchema = async () => {
  const toJsonSchema = (schema: unknown) =>
    (schema as Type).toJsonSchema()

  // eslint-disable-next-line @masknet/type-prefer-return-type-annotation
  return toJsonSchema as (schema: unknown) => JSONSchema7
}
