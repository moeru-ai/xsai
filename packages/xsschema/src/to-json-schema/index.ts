import type { StandardSchemaV1 } from '@standard-schema/spec'
import type { JSONSchema7 } from 'json-schema'

export const toJsonSchema = async (schema: StandardSchemaV1): Promise<JSONSchema7> => {
  switch (schema['~standard'].vendor) {
    case 'valibot':
      // eslint-disable-next-line ts/no-unsafe-argument
      return (await import('./valibot')).toJsonSchema(schema as any)
    case 'zod':
      // eslint-disable-next-line ts/no-unsafe-argument, @masknet/type-prefer-return-type-annotation
      return (await import('./zod')).toJsonSchema(schema as any) as JSONSchema7
    default:
      throw new Error(`xsschema: Unsupported schema vendor ${schema['~standard'].vendor}`)
  }
}
