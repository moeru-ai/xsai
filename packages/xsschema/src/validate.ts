import type { StandardSchemaV1 } from '@standard-schema/spec'

/** @see {@link https://github.com/standard-schema/standard-schema#how-do-i-accept-standard-schemas-in-my-library} */
export const validate = async <T extends StandardSchemaV1>(schema: T, input: StandardSchemaV1.InferInput<T>): Promise<StandardSchemaV1.InferOutput<T>> => {
  let result = schema['~standard'].validate(input)
  if (result instanceof Promise)
    result = await result

  if (result.issues)
    throw new Error(JSON.stringify(result.issues, null, 2))

  return (result as StandardSchemaV1.SuccessResult<T>).value
}
