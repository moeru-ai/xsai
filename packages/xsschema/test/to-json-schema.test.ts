import { type } from 'arktype'
import { describe, expect, it } from 'vitest'

import { toJsonSchema } from '../src'

describe('toJsonSchema', () => {
  it('arktype', async () => {
    const schema = type({
      myString: 'string',
      myUnion: 'number | boolean',
    }).describe('My neat object schema')

    const jsonSchema = await toJsonSchema(schema)

    expect(jsonSchema).toMatchSnapshot()
  })
})
