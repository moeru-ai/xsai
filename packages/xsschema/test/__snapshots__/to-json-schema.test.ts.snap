// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`toJsonSchema > arktype 1`] = `
{
  "description": "My neat object schema",
  "properties": {
    "myString": {
      "type": "string",
    },
    "myUnion": {
      "anyOf": [
        {
          "type": "number",
        },
        {
          "type": "boolean",
        },
      ],
    },
  },
  "required": [
    "myString",
    "myUnion",
  ],
  "type": "object",
}
`;
